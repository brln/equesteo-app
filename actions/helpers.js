import BackgroundGeolocation from 'react-native-mauron85-background-geolocation'
import firebase from 'react-native-firebase'

import {
  dismissError,
  errorOccurred,
  loadCurrentRideState,
  loadLocalState,
  localDataLoaded,
  saveUserID,
  setAwaitingPasswordChange,
  setDoingInitialLoad,
} from './standard'

import {
  getFCMToken,
  startListeningFCMTokenRefresh,
  startListeningFCM,
  setDistributionOnServer,
  switchRoot,
} from './functional'
import { logError, logInfo } from '../helpers'
import { FEED } from '../screens'
import { LocalStorage, PouchCouch } from '../services'

import { setUserContext } from "../services/Sentry"

export function loginAndSync(loginFunc, loginArgs, dispatch) {
  loginFunc(...loginArgs).then(resp => {
    // @TODO: figure out why followers have _id here!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1
    const userID = resp.id
    const following = resp.following
    const followers = resp.followers

    dispatch(dismissError())
    // dispatch(setAwaitingPasswordChange(true))
    dispatch(saveUserID(userID))
    // setUserContext(userID)
    // dispatch(getFCMToken())
    // dispatch(startListeningFCMTokenRefresh())
    // dispatch(setDistributionOnServer())
    dispatch(setDoingInitialLoad(true))
    return PouchCouch.localReplicateDB('all', [...following, userID], followers)
  }).then(() => {
    return PouchCouch.localLoad()
  }).then(localData => {
    dispatch(localDataLoaded(localData))
    dispatch(setDoingInitialLoad(false))
    dispatch(switchRoot(FEED))
    dispatch(startListeningFCM())
  }).catch(e => {
    logDebug(JSON.stringify(e), 'this error specifically')
    dispatch(errorOccurred(e.message))
  })
}

export function tryToLoadStateFromDisk (dispatch) {
  return Promise.all([
    LocalStorage.loadLocalState(),
    LocalStorage.loadCurrentRideState()
  ]).then(([localState, currentRideState]) => {
    if (localState) {
      dispatch(loadLocalState(localState))
    } else {
      logInfo('no cached local state found')
    }

    if (currentRideState) {
      dispatch(loadCurrentRideState(currentRideState))
    } else {
      logInfo('no cached current ride state found')
    }
  })
}

export function configureBackgroundGeolocation () {
  return new Promise((res, rej) => {
    BackgroundGeolocation.configure(
      {
        desiredAccuracy: BackgroundGeolocation.HIGH_ACCURACY,
        locationProvider: BackgroundGeolocation.RAW_PROVIDER,
        distanceFilter: 0,
        maxLocations: 10,
        interval: 0,
        notificationTitle: 'You\'re out on a ride.',
        notificationText: 'Tap here to see your progress.',
      },
      () => {
        logInfo('Background geolocation configured')
        res()
      },
      (e) => {
        rej(e)
      },
    )
  })
}

export function stopListeningFCM () {
  // maybe delete token on server here?
  return firebase.iid().deleteToken('373350399276', 'GCM').then(() => {
    firebase.messaging().onTokenRefresh(() => {})
  })
}
