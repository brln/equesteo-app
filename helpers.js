const toRad = (deg) => {
  return deg * Math.PI / 180;
}

const toDeg = (rad) => {
  return rad * 180 / Math.PI
}


export const haversine = (lat1, lon1, lat2, lon2) => {
  const R = 3959; // mi
  const x1 = lat2 - lat1
  const dLat = toRad(x1)
  const x2 = lon2 - lon1
  const dLon = toRad(x2)
  const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon/2) * Math.sin(dLon/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export const bearing = (lat1,lng1,lat2,lng2) => {
  let dLon = toRad(lng2-lng1);
  let y = Math.sin(dLon) * Math.cos(toRad(lat2));
  let x = Math.cos(toRad(lat1))*Math.sin(toRad(lat2)) - Math.sin(toRad(lat1))*Math.cos(toRad(lat2))*Math.cos(dLon);
  let brng = toDeg(Math.atan2(y, x));
  return ((brng + 360) % 360);
}

export const rideCoordsToMapCoords = (rideCoords) => {
  const sorted = [...rideCoords].sort((a, b) => {
    return new Date(a.timestamp) - new Date(b.timestamp);
  })

  return sorted.map((apiCoord) => {
    return {
      latitude: parseFloat(apiCoord.latitude),
      longitude: parseFloat(apiCoord.longitude),
    }
  })
}

export const unixTimeNow = () => {
  return Math.floor(new Date().getTime())
}

export const profilePhotoURL = (id) => {
  return `https://s3.amazonaws.com/equesteo-profile-photos/full_size/${id}.jpeg`
}

// from https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
export function generateUUID () { // Public Domain/MIT
    let d = new Date().getTime();
    if (typeof performance !== 'undefined' && typeof performance.now === 'function'){
        d += performance.now(); //use high-precision timer if available
    }
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        var r = (d + Math.random() * 16) % 16 | 0;
        d = Math.floor(d / 16);
        return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
    });
}

function urlParams (params) {
  return Object.keys(params).map(k => k + '=' + encodeURIComponent(params[k])).join('&')
}

export function staticMap (ride) {
  const STATIC_MAPS_API_KEY = ' AIzaSyBhUmpq-7uQ2JaqtrHO3hpfeFHynVpo8xQ'
  const ROOT_URL = 'https://maps.googleapis.com/maps/api/staticmap?'
  const queryStringParams = {
      size: '400x400',
      format: 'png',
      maptype: 'terrain',
  }
  const pathStyle = 'color:0xff0000ff|weight:5'
  let pathCoords = ''
  for (let coordinate of ride.rideCoordinates) {
    pathCoords += `|${coordinate.latitude},${coordinate.longitude}`
  }

  queryStringParams['path'] = pathStyle + pathCoords
  queryStringParams['key'] = STATIC_MAPS_API_KEY
  const queryString = urlParams(queryStringParams)
  return ROOT_URL + queryString
}
