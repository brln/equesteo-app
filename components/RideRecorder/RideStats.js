import React, { PureComponent } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import memoizeOne from 'memoize-one';

import { darkGrey, lightGrey } from '../../colors'
import { metersToFeet } from '../../helpers'

const initialState = {
  starting: null,
  elapsedTime: undefined
}

export default class RideStats extends PureComponent {
  constructor () {
    super()
    this.avgSpeed = this.avgSpeed.bind(this)
    this.state = { ...initialState }
    this.elapsedAsString = this.elapsedAsString.bind(this)
    this.currentAltitude = this.currentAltitude.bind(this)
    this.currentSpeed = this.currentSpeed.bind(this)
    this.timeToString = this.timeToString.bind(this)

    this.memoAvgSpeed = memoizeOne(this.avgSpeed)
    this.memoCurrentAltitude = memoizeOne(this.currentAltitude)
    this.memoCurrentSpeed = memoizeOne(this.currentSpeed)
    this.memoTimeToString = memoizeOne(this.timeToString)
  }

  avgSpeed (time, distance) {
    const elapsed = time
    let hours = 0
    let minutes = 0
    let seconds = 0
    if (elapsed > 0 && this.props.distance) {
      hours = elapsed.getUTCHours()
      minutes = elapsed.getUTCMinutes()
      seconds = elapsed.getUTCSeconds()

      const totalHours = hours + (minutes / 60) + (seconds / 60 / 60)
      const milesPerHour = distance / totalHours
      return milesPerHour.toFixed(1).toString()
    } else {
      return '0.0'
    }
  }

  currentSpeed (rideCoords) {
    const last = rideCoords.get(-1)
    if (last) {
      const found = last.get('speed')
      if (found || found === 0) {
        return (last.get('speed') * 2.236936).toFixed(1)
      } else {
        return '-'
      }
    } else {
      return '0.0'
    }
  }

  currentAltitude (rideCoords, last) {
    if (last) {
      const found = last.get('elevation')
      if (found) {
        return Math.round(metersToFeet(last.get('elevation')))
      } else {
        return '-'
      }
    } else {
      return '0'
    }
  }

  componentDidMount() {
    if (this.props.startTime && !this.state.startTime) {
      this.setState({
        startTime: this.props.startTime,
        elapsedTime: new Date(new Date() - this.props.startTime)
      })
      this.renderTimer = setInterval(() => {
        this.setState({
          elapsedTime: new Date(new Date() - this.state.startTime)
        })
      }, 100)
    }
  }

  componentWillUnmount() {
    this.setState({ ...initialState })
    clearInterval(this.renderTimer)
  }

  leftpad(num) {
    const str = num.toString()
    const pad = "00"
    return pad.substring(0, pad.length - str.length) + str
  }

  timeToString (elapsed, elapsedSeconds) {
    const hours = this.leftpad(elapsed.getUTCHours())
    const minutes = this.leftpad(elapsed.getUTCMinutes())
    const seconds = this.leftpad(elapsedSeconds)
    return `${hours}:${minutes}:${seconds}`
  }

  elapsedAsString () {
    const elapsed = this.state.elapsedTime
    if (elapsed) {
      return this.memoTimeToString(elapsed, elapsed.getUTCSeconds())
    }
  }

  render () {
    return (
      <View style={styles.rideStats}>
        <View style={{flex: 1}}>
          <View style={[styles.statBox, {borderTopWidth: 2, borderBottomWidth: 1, borderRightWidth: 1, borderLeftWidth: 2}]}>
            <Text style={styles.statName}>Distance</Text>
            <Text style={styles.statFont}>
              {this.props.distance.toFixed(2)} <Text style={styles.unitsFont}>mi</Text>
            </Text>
          </View>
          <View style={[styles.statBox, {borderTopWidth: 1, borderBottomWidth: 2, borderRightWidth: 1, borderLeftWidth: 2}]}>
            <Text style={styles.statName}>Current Speed</Text>
            <Text style={styles.statFont}>{this.memoCurrentSpeed(this.props.rideCoords)} <Text style={styles.unitsFont}>mi/h</Text></Text>
          </View>
          <View style={[styles.statBox, {borderTopWidth: 1, borderBottomWidth: 2, borderRightWidth: 1, borderLeftWidth: 2}]}>
            <Text style={styles.statName}>Altitude</Text>
            <Text style={styles.statFont}>{this.memoCurrentAltitude(this.props.rideCoords, this.props.lastElevation)} <Text style={styles.unitsFont}>ft</Text></Text>
          </View>
        </View>
        <View style={{flex: 1}}>
          <View style={[styles.statBox, {borderTopWidth: 2, borderBottomWidth: 1, borderRightWidth: 2, borderLeftWidth: 1}]}>
            <Text style={styles.statName}>Total Time</Text>
            <Text style={styles.statFont}>{this.elapsedAsString()}</Text>
          </View>
          <View style={[styles.statBox, {borderTopWidth: 1, borderBottomWidth: 2, borderRightWidth: 2, borderLeftWidth: 1}]}>
            <Text style={styles.statName}>Avg. Speed</Text>
            <Text style={styles.statFont}>
              {this.memoAvgSpeed(this.state.elapsedTime, this.props.distance)} <Text style={styles.unitsFont}>mi/h</Text>
            </Text>
          </View>
          <View style={[styles.statBox, {borderTopWidth: 1, borderBottomWidth: 2, borderRightWidth: 2, borderLeftWidth: 1}]}>
            <Text style={styles.statName}>Elevation Gain</Text>
            <Text style={styles.statFont}>
              {Math.round(metersToFeet(this.props.currentRideElevations.get('elevationGain')))} <Text style={styles.unitsFont}>ft</Text>
            </Text>
          </View>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  rideStats: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  statFont: {
    color: 'black',
    fontSize: 35,
    textAlign: 'center',
  },
  statName: {
    paddingLeft: 5,
    color: darkGrey,
    fontSize: 14,
  },
  statBox: {
    flex: 1,
    borderColor: lightGrey
  },
  unitsFont: {
    color: 'black',
    fontSize: 25
  }
});
