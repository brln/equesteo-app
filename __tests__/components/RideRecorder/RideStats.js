import { Map } from 'immutable'
import React from 'react';
import renderer from 'react-test-renderer';

import { unixTimeNow } from '../../../helpers'
import RideStats from '../../../components/RideRecorder/RideStats'

test('renders nothing when not visible', () => {
  const currentRide = Map({
    startTime: unixTimeNow()
  })
  const tree = renderer.create(
    <RideStats
      currentRide={currentRide}
      visible={false}
  />).toJSON();
  expect(tree).toBeNull()
});

test('renders', () => {
  const currentRide = Map({
    startTime: unixTimeNow(),
    distance: 32
  })
  const currentRideElevations = Map({
    elevationGain: 20
  })

  const tree = renderer.create(
    <RideStats
      currentRide={currentRide}
      currentRideElevations={currentRideElevations}
      visible={true}
    />).toJSON();
  expect(tree).toMatchSnapshot()
});
