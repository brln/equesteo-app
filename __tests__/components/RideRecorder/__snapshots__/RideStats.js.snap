// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 3,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 1,
            "borderLeftWidth": 2,
            "borderRightWidth": 1,
            "borderTopWidth": 2,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Distance
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          32.00
           
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
              }
            }
          >
            mi
          </Text>
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 2,
            "borderLeftWidth": 2,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Current Speed
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          0.0
           
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
              }
            }
          >
            mi/h
          </Text>
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 2,
            "borderLeftWidth": 2,
            "borderRightWidth": 1,
            "borderTopWidth": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Altitude
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          0
           
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
              }
            }
          >
            ft
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 2,
            "borderTopWidth": 2,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Total Time
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 2,
            "borderLeftWidth": 1,
            "borderRightWidth": 2,
            "borderTopWidth": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Avg. Speed
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          0.0
           
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
              }
            }
          >
            mi/h
          </Text>
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "borderColor": "#e8eae8",
            "flex": 1,
          },
          Object {
            "borderBottomWidth": 2,
            "borderLeftWidth": 1,
            "borderRightWidth": 2,
            "borderTopWidth": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#9b9d9b",
            "fontSize": 14,
            "paddingLeft": 5,
          }
        }
      >
        Elevation Gain
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 35,
              "textAlign": "center",
            }
          }
        >
          66
           
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
              }
            }
          >
            ft
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
